wandb/summary.py

```diff
@@ -1,61 +1,137 @@
+                return {key: self._transform(k+"/"+key, value, write=False) for (key, value) in v.items()}
         else:
             return v
 
     def __getitem__(self, k):
-        return self._summary[k]
+        return self._transform(k, self._summary[k], write=False)
 
     def __setitem__(self, k, v):
-        self._summary[k] = self._transform(v)
+        self._summary[k] = self._transform(k, v)
         self._write()
 
     def __setattr__(self, k, v):
         if k.startswith("_"):
             super(Summary, self).__setattr__(k, v)
         else:
-            self._summary[k] = self._transform(v)
+            self._summary[k] = self._transform(k, v)
             self._write()
 
     def __getattr__(self, k):
         if k.startswith("_"):
-            super(Summary, self).__getattr__(k)
+            return super(Summary, self).__getattr__(k)
         else:
-            return self._summary[k]
+            return self._transform(k, self._summary[k], write=False)
 
     def __delitem__(self, k):
+        val = self._summary[k]
+        if isinstance(val, dict) and val.get("_type") in H5_TYPES:
+            if not self._h5:
+                wandb.termerror("Deleting tensors in summary requires h5py")
+            else:
+                del self._h5["summary/" + k]
+                self._h5.flush()
         del self._summary[k]
         self._write()
 
     def __repr__(self):
         return str(self._summary)
 
+    def keys(self):
+        return self._summary.keys()
+
     def get(self, k, default=None):
         return self._summary.get(k, default)
 
+    def write_h5(self, key, val):
+        # ensure the file is open
+        self.open_h5()
+
+        if not self._h5:
+            wandb.termerror("Storing tensors in summary requires h5py")
+        else:
+            try:
+                del self._h5["summary/" + key]
+            except KeyError:
+                pass
+            self._h5["summary/" + key] = val
+            self._h5.flush()
+
+    def read_h5(self, key, val):
+        # ensure the file is open
+        self.open_h5()
+
+        if not self._h5:
+            wandb.termerror("Reading tensors from summary requires h5py")
+        else:
+            return self._h5["summary/" + key]
+
+    def open_h5(self):
+        if not self._h5 and h5py:
+            self._h5 = h5py.File(self._h5_path, 'a', libver='latest')
+
+    def convert_json(self, obj=None, root_path=[]):
+        """Convert obj to json, summarizing larger arrays in JSON and storing them in h5."""
+        res = {}
+        obj = obj or self._summary
+        for key, value in six.iteritems(obj):
+            path = ".".join(root_path + [key])
+            if isinstance(value, dict):
+                res[key], converted, transformed = util.json_friendly(
+                    self.convert_json(value, root_path + [key]))
+            else:
+                res[key], converted, transformed = util.json_friendly(value)
+                if transformed:
```

Comment:
Here I'm actually writing the key in H5 as "summary/key.nested".  I did this because h5 was being a funky about having nested groups...